# S&P 2019 Data Wrangling

**by Marc Angelo Acebedo**

## Table of Contents
- [Introduction](#introduction)
- [Data Wrangling](#wrangling)
    - [Gather](#gather)
    - [Assess](#assess)
    - [Clean](#clean)
    - [Store](#store)

## I) Introduction <a id = "introduction">

**Broad question:** How do price forecasts for each firm in the S&P 2019 Index compare to their corresponding actual prices?

**Approach:** Analyze difference in means between average forecast EPS and average actual EPS for each firm.

I will be analyzing quarterly price returns within the past 20 years for the firms present in the S&P 500 2019 Index.

> At first, I wanted to analyze the forecasted vs. actual price earnings of the S&P in its entirety for the past 20 years. However, considering that firms continuously enter and leave stock indices every year, there would be varying levels of inconsistencies and marginal errors when comparing annual S&P returns alone. To combat this problem, I have isolated these two approaches:
- Analyze the historical earnings of *only* the firms present in the S&P 2019 Index
- Keep track of all firms that were present in the S&P for the past 20 years. Keep track of how many times each firm appeared in the Index and for those with the least count, analyze them individually on how they differ from the firms that stayed for longer.

[TK] HEre is a breakdown of my final clean CSV's features, TK.csv.

## II) Data Wrangling <a id="wrangling"></a>

To gather the data depicted under the `./data` folder, I used Bloomberg Excel functions.

### A) Gather <a id = "gather"></a>
> **APPROACH 1:** Focus on the firms that appear in the 2019 S&P Index and analyze their forecasted vs. actual price earnings for the last 20 years.

To ensure consistency in analysis among multiple firms, I divide both the forecasted and actual price earning dates by *calendar period* instead of fiscal period. This is because fiscal period differs by firm whereas calendar period is consistent by dates. 

#### Through the Bloomberg Excel functions, I gathered four datasets with different purposes:

- historical forecasted EPS
- historical actual EPS
- historical actual EOD price
- historical forecasted EPS relying on terms

---
Before delving into the data, let's define what above terms:

**EPS**

> EPS stands for ***Estimated Price Earnings.*** The formal definition of EPS given by Investopedia is this:

Earnings per share is the portion of a company's profit that is allocated to each outstanding share of a common stock, serving as an indicator of the company's financial health.

In other words, the EPS is a portion of the company's **net income** after all of their dividends are paid off. Dividends are profits that are paid out to shareholders of the company. EPS is one of the most useful and valuable financial measurements because they ***determine a stock's worth.*** The higher the stock, the more the company can pay out dividends to its shareholders, and the more net profit they are determined to generate.

$$ EPS = \frac{Net Income - Preferred Dividends}{Weighted Average Common Shares Outstanding}\$$

**EOD**

> EOD stands for the ***End of Day*** price. For any given day, the EOD marks the ***price at which the stock was valued*** at the end of the day's trading period.

In [1]:
PATH = './data/'
PATH_CLEAN = './data/clean/'
PATH_CLEAN_AVGS = './data/clean/averages/'
PATH_CLEAN_AVGS_SUB = './data/clean/averages/components/'

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

from functools import reduce

### Import Data <a id = "import"></a>

### Let's summarize the contents of the following DataFrames as we import them.
> All DataFrames consist of the 505 firms found in the 2019 S&P Index with EPS and EOD data encompassing 20 years: from January 1999 until the December 2019.

**Historic forecasted EPS**
> According to Investopedia, consensus estimates is normally an average or median of all the forecasts from individual analysts tracking a particular stock. In this case, the consensus estimate is for ***EPS for each firm present in the index as of 2019.*** Forecasted EPS is calculated by ***quarterly earnings,*** usually by each firm's fiscal period. Estimates of quarterly earnings are published at the beginning of each quarterly period.

In [2]:
#historic forecasted EPS 
eps_fc = pd.read_csv(PATH + 'sp-eps-fc.csv')

In [3]:
eps_fc.head()

Term Forecasted  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
0          1999Q1          NaN            NaN            NaN           0.025   
1          1999Q2          NaN            NaN            NaN           0.020   
2          1999Q3          NaN            NaN            NaN           0.023   
3          1999Q4          NaN            NaN            NaN           0.016   
4            00Q1          NaN            NaN            NaN           0.032   

   ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
0             NaN            NaN           -0.10          0.423   
1             NaN            NaN           -0.13          0.420   
2             NaN            NaN           -0.11          0.381   
3             NaN            NaN             NaN          0.432   
4             NaN            NaN             NaN          0.441   

   ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
0            NaN  ...          0.456           0.090          0.239   
1            NaN  ...          0.213           0.090          0.264   
2            NaN  ...          0.767           0.099          0.300   
3            NaN  ...          0.432           0.124          0.386   
4            NaN  ...          0.283           0.143          0.436   

   XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  \
0           0.123            NaN            NaN          0.123            NaN   
1           0.135            NaN            NaN          0.150            NaN   
2           0.129            NaN            NaN          0.171            NaN   
3           0.178            NaN            NaN          0.164            NaN   
4           0.138            NaN            NaN          0.138            NaN   

   ZION UW Equity  ZTS UN Equity  
0           0.622            NaN  
1           0.661            NaN  
2           0.693            NaN  
3           0.667            NaN  
4           0.668            NaN  

[5 rows x 506 columns]

**Historic actual EPS**
> Unlike forecasted EPS, actual EPS are the real numbers denoting Earnings-per-Share for a singular firm. Historic actual EPS will be compared to forecasted EPS to draw correlations and comparisons.

In [4]:
#historic actual EPS
eps_act = pd.read_csv(PATH + 'sp-eps-act.csv')

In [5]:
eps_act.head()

Quarter  Year  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
0      Q1  1999         0.16           0.99            NaN        0.040000   
1      Q2  1999         0.35           1.76            NaN        0.035357   
2      Q3  1999         0.30           1.86            NaN        0.050357   
3      Q4  1999         0.32           1.89            NaN        0.024643   
4      Q1  2000         0.29           0.89            NaN        0.040585   

   ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  ...  \
0             NaN         0.0925          -0.030           0.44  ...   
1             NaN         0.1000          -0.145           0.42  ...   
2             NaN         0.1075          -0.135           0.30  ...   
3             NaN         0.0375          -0.080           0.43  ...   
4             NaN         0.1050          -0.045           0.45  ...   

   XEL UW Equity  XLNX UW Equity  XOM UN Equity  XRAY UW Equity  \
0           0.34          0.0000          0.210        0.123333   
1           0.06          0.0975          0.285        0.133333   
2           0.63          0.0325          0.315        0.130000   
3           0.43          0.1300          0.600        0.180000   
4           0.45          0.1650          0.500        0.140000   

   XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  ZION UW Equity  \
0           1.64            NaN         0.1725            NaN            0.56   
1           2.28            NaN         0.2900            NaN            0.60   
2           1.96            NaN         0.3200            NaN            0.64   
3           1.72            NaN         0.2375            NaN            0.49   
4          -1.48            NaN         0.2025            NaN           -0.33   

   ZTS UN Equity  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 507 columns]

**Historic actual EOD**
> Though this is not directly related to EPS data, EOD would be an interesting measure to use when generating intriguing visualizations and analyses. Who knows what visuals and conclusions I would arrive to with this measure. 

In [6]:
#historic actual EOD
eod_act = pd.read_csv(PATH + 'sp-eod-act.csv')

In [7]:
eod_act.head()

date  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
0   3/31/1999          NaN            NaN            NaN          1.2835   
1   6/30/1999          NaN            NaN            NaN          1.6540   
2   9/30/1999          NaN            NaN            NaN          2.2612   
3  12/31/1999      52.0909            NaN            NaN          3.6719   
4   3/31/2000      70.0721            NaN            NaN          4.8504   

   ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
0             NaN         8.2934           6.250        20.9504   
1             NaN         6.1859           6.875        20.3630   
2             NaN         5.7462           7.750        16.4471   
3             NaN         3.6843          18.375        16.2513   
4             NaN         3.6388          20.250        15.7478   

   ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
0            NaN  ...            NaN         20.2813        35.2813   
1            NaN  ...            NaN         28.6250        38.5625   
2            NaN  ...            NaN         32.7656        37.9688   
3            NaN  ...            NaN         45.4688        40.2813   
4            NaN  ...            NaN         82.8125        38.9063   

   XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  \
0          7.7500       140.6214            NaN        12.6284            NaN   
1          9.6250       155.6057            NaN         9.7297            NaN   
2          7.5833       110.4883            NaN         7.3591            NaN   
3          7.8750        59.7723            NaN         6.9434            NaN   
4          9.4583        68.4994            NaN         5.5839            NaN   

   ZION UW Equity  ZTS UN Equity  
0         66.5000            NaN  
1         63.5000            NaN  
2         55.1250            NaN  
3         59.1875            NaN  
4         41.6250            NaN  

[5 rows x 506 columns]

**Historic forecasted EPS 3 months prior**
> Instead of using forecast data collected at the beginning of the fiscal period, this feature contains EPS data projected 3 months before the current fiscal period. This is an interesting metric to see how differently forecasters make their predictions at different times.

In [8]:
#historic forecasted EPS 3-months prior
eps_fc_terms = pd.read_csv(PATH + 'sp-eps-fc-terms.csv')

In [9]:
eps_fc_terms.head()

Forecast Made Term Forecasted  A UN Equity  AAL UW Equity  AAP UN Equity  \
0     10/1/1999            00Q1          NaN            NaN            NaN   
1      1/1/2000            00Q2          NaN            NaN            NaN   
2      4/1/2000            00Q3          NaN            NaN            NaN   
3      7/1/2000            00Q4          NaN            NaN            NaN   
4      9/1/2000            01Q1          NaN            NaN            NaN   

   AAPL UW Equity  ABBV UN Equity  ABC UN Equity  ABMD UW Equity  \
0           0.030             NaN            NaN             NaN   
1           0.026             NaN            NaN             NaN   
2           0.029             NaN            NaN             NaN   
3           0.032             NaN            NaN           -0.18   
4           0.041             NaN            NaN           -0.09   

   ABT UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
0          0.466  ...            NaN           0.143          0.398   
1          0.437  ...            NaN           0.168          0.378   
2          0.414  ...           0.84           0.189          0.425   
3          0.477  ...           0.53           0.211          0.487   
4          0.495  ...           0.40           0.259          0.475   

   XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  \
0           0.140            NaN            NaN          0.144            NaN   
1           0.153            NaN            NaN          0.184            NaN   
2           0.149            NaN            NaN          0.218            NaN   
3           0.200            NaN            NaN          0.244            NaN   
4           0.158            NaN            NaN          0.178            NaN   

   ZION UW Equity  ZTS UN Equity  
0           0.777            NaN  
1           0.790            NaN  
2           0.803            NaN  
3           0.777            NaN  
4           0.750            NaN  

[5 rows x 507 columns]

### B) Assess <a id = "assess"></a>

> The following DataFrames contain data for each firm across various dates. To account for all firm averages, my goal is to generate a CSV file where each row contains the firm average, with the features as columns.

In [10]:
dict_dfs = {'eps_fcast' : eps_fc,
           'eps_actual' : eps_act,
           'eod_actual' : eod_act,
           'eps_fcast_terms' : eps_fc_terms}

### Check Missing Data

In [11]:
for key, df in dict_dfs.items():
    print(key, df.shape)

eps_fcast (84, 506)
eps_actual (84, 507)
eod_actual (84, 506)
eps_fcast_terms (80, 507)


***We need to make sure the number of firms in each DataFrame is consistent.***

---

**Observation 1:** for `eps_fcast`, there are 505 firms encompassing 84 quarterly fiscal periods since 1999.

> There are 506 columns: 1 column being `Term Forecast`, the rest firm names.

**Observation 2:** for `eps_actual`, there are 505 firms encompassing 84 quarterly fiscal periods since 1999.

> There are 507 columns: 2 columns being `Quarter` and `Year`, the rest firm names.

**Observation 3:** for `eod_actual`, there are 505 firms encompassing 84 quarterly calendar periods since 1999.

> There are 506 columns: 1 column being `date`, the rest firm names.

**Observation 4:** for `eps_fcast_terms`, there are 505 firms encompassing 80 quarterly calendar periods since 1999.

> There are 507 columns: 2 columns being `Forecast Made` and `Term Forecasted`, the rest firm names.
 
Since there are only 80 quarterly calendar periods, that ***implies an entire year is missing.***

**Observation 5:** For `eps_fcast`, `eps_actual`, and `eod_actual`, since there are 4 quarters in a year, 84 quarterly forecast periods equate to 21 years. This is correct since we are analyzing the years from 1999 until the end of 2019.

### Most importantly, the number of firms across all DataFrames is consistent.

----

***Check which year is missing from `eps_fcast_terms`***.

> Each **Term Forecasted** entry under `eps_fcast_terms` records the year with 2 digits, so Quarter 1 of the year 2000 becomes 00Q1.
- isolate the first 2 characters to get the year
- join a '20' in front of the string so 00 becomes 2000
- list the number of unique values.

In [12]:
#iterate over all years in eps_fcast_terms, append '20' in front of string
list(map(lambda x: '20' + x, eps_fc_terms['Term Forecasted'].str[:2].unique()))

['2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013',
 '2014',
 '2015',
 '2016',
 '2017',
 '2018',
 '2019']

**Observation**: The year 1999 is missing from `eps_fc_terms`. This makes sense because the start of the forecasting period would be in the last quarter of 1999, which is October. 

---

### Check Nulls

In [13]:
for key, df in dict_dfs.items():
    #display (row, column) per DataFrame
    print('# NaN in {}: {}'.format(key, df.isna().sum().sum()))

# NaN in eps_fcast: 7055
# NaN in eps_actual: 5021
# NaN in eod_actual: 6921
# NaN in eps_fcast_terms: 7080


**Observation:** All four DataSets contain null values.
> In order to combat this, we'll have to look at both the **number of rows** and **number of columns** with missing data, separately. This way, we can isolate which firms and/or time periods contain complete or incomplete data.

In [14]:
#check rows for missing data
for key, df in dict_dfs.items():
    num_rows_missing = df.isna().sum().max()
    print('{} has {} time periods containing missing data, out of {} total rows.'.format(key, num_rows_missing, df.shape[0]))

eps_fcast has 82 time periods containing missing data, out of 84 total rows.
eps_actual has 84 time periods containing missing data, out of 84 total rows.
eod_actual has 83 time periods containing missing data, out of 84 total rows.
eps_fcast_terms has 80 time periods containing missing data, out of 80 total rows.


In [15]:
len(eps_fc.columns[eps_fc.isnull().any()])

254

In [16]:
#check columns for missing data
for key, df in dict_dfs.items():
    cols_missing = df.columns[df.isnull().any()]
    num_cols_missing = len(cols_missing)
    print('{} has {} firms containing missing data, out of {} total columns.'
         .format(key, num_cols_missing, df.shape[1]))

eps_fcast has 254 firms containing missing data, out of 506 total columns.
eps_actual has 432 firms containing missing data, out of 507 total columns.
eod_actual has 505 firms containing missing data, out of 506 total columns.
eps_fcast_terms has 340 firms containing missing data, out of 507 total columns.


**Observation 1:** The only two datasets with incomplete data for all time periods are **actual EPS** and **forecasted EPS 3 months prior.**

> To address this problem, it'd be helpful to isolate the time period ranges for the datasets with incomplete data by row, **forecasted EPS** and **actual EOD price.**

**Observation 2:** For all datasets, all firms contain incomplete data across all time periods.
> This is expected, as analyzing financial history spanning over 20 years will naturally be rife with missing and inaccurate data. The ***good news is that `eps_fcast`, `eps_actual`, and `eps_fcast_terms` are the most complete, while `eod_actual` contains the most amount of missing data.***


**Moving forward, we need to make sure that these inconsistencies won't clash with our analysis.**
> **My approach:** instead of looking at rows and columns ***with*** missing data, we'll be looking at rows and columns that ***are all missing data.***

I figured that if there is some missing data here and there scattered throughout the matrix, then that should not skew our analysis too much.

However, if there a significant amount of rows/columns that are entirely empty, then we ***might have to get ready to drop some dates and firms from our data overall.***

In [17]:
#check for empty rows, return False if row contains at least one non-null value, True if all are null
for key, df in dict_dfs.items():
    cols_check = df.columns
    num_empty_rows = (df[cols_check].isnull().apply(lambda x: all(x), axis = 1)).value_counts()
    print(key, '\n', num_empty_rows, '\n----')

eps_fcast 
 False    84
dtype: int64 
----
eps_actual 
 False    84
dtype: int64 
----
eod_actual 
 False    84
dtype: int64 
----
eps_fcast_terms 
 False    80
dtype: int64 
----


**Observation:** All datasets do not contain empty rows.
> This is good news, since we can rely on the firms' averages per row instead of having to drop or limit time periods.

In [18]:
#check for empty columns, return False if column contains at least one non-null value, True if all are null
for key, df in dict_dfs.items():
    cols_check = df.columns
    num_empty_cols = df[cols_check].isnull().apply(lambda x: all(x), axis = 0).value_counts()
    print(key, '\n', num_empty_cols)

eps_fcast 
 False    506
dtype: int64
eps_actual 
 False    506
True       1
dtype: int64
eod_actual 
 False    506
dtype: int64
eps_fcast_terms 
 False    506
True       1
dtype: int64


**Observation:** `eps_actual` and `eps_fcast_terms` are the only datasets that have an empty column.
> Let's isolate and look at the singular empty column for both DataFrames.

In [19]:
#create function to return an array of column names containing empty data
def comb_cols(df):
    empty_cols = []
    for column in df:
        if df[column].isnull().all():
            empty_cols.append(column)
            
    return empty_cols

In [20]:
#comb datasets for empty columns
print('In eps_act, the firm {} has no data.'.format(comb_cols(eps_act)))
print('In eps_fc_terms, the firm {} has no data.'.format(comb_cols(eps_fc_terms)))

In eps_act, the firm ['AMCR UN Equity'] has no data.
In eps_fc_terms, the firm ['AMCR UN Equity'] has no data.


**Observation:** The same firm in both datasets is empty of data.
> Though this is an annoying error to deal with, it still is to our advantage that both datasets ***share one firm*** in common for missing data. This way, we don't have to worry about dropping two entire firms.

---

### Check Duplicate Data

In [21]:
#check for duplicate data across all rows and columns
for key, df in dict_dfs.items():
    print(key, df.duplicated().sum())

eps_fcast 0
eps_actual 0
eod_actual 0
eps_fcast_terms 0


**Observation:** For all datasets, there are ***no duplicate data.*** This is good news!

**Next, I will check for duplicated firm names.** Although the presence of duplicated firm names will inherently imply duplicated data, sometimes data gets dispersed in weird, unexpected ways, especially when dealing with large datasets.

In [22]:
#check for duplicated firm names
for key, df in dict_dfs.items():
    print(key, df.columns.duplicated().sum())

eps_fcast 0
eps_actual 0
eod_actual 0
eps_fcast_terms 0


**Observation:** For all datasets, there are ***no duplicate firm names.*** This is also good news.

### There is no need to dedupe our data during the cleaning stage.

---

### Check Data Types

> Make sure all numerical data types are consistent.
- EPS and EOD values must all be **floats**
- Dates should all be **objects** (for now)

- `eps_fc` contains ***1 date field,***, so we should expect ***1 object type.***
- `eps_act` contains ***2 date fields,*** so we should expect ***2 object types.***
- `eod_act` contains ***1 date field,*** so we should expect ***1 object type.***
- `eps_fc_terms` contains ***2 date fields,*** so we should expect ***2 object types.***

> And everything else should be a ***float*** type.

**Examine each dataset's overall data types per column.**

In [23]:
for key, df in dict_dfs.items():
    print(key,'\n', df.dtypes.value_counts(), '\n--------')

eps_fcast 
 float64    505
object       1
dtype: int64 
--------
eps_actual 
 float64    499
object       7
int64        1
dtype: int64 
--------
eod_actual 
 float64    505
object       1
dtype: int64 
--------
eps_fcast_terms 
 float64    505
object       2
dtype: int64 
--------


**Observation 1:** `eps_fc`, `eod_act`, and `eps_fc_terms` are all consistent with the expected number of object types.

> But we'll still have to double check these facts later.

**Observation 2:** `eps_act` contains 5 more object columns than expected.

---

***Isolate 'Object' Columns under `eps_act`***

In [24]:
eps_act.select_dtypes(include = 'object').head()

Quarter            BRK/B UN Equity              FOX UW Equity  \
0      Q1  #N/A Field Not Applicable  #N/A Field Not Applicable   
1      Q2  #N/A Field Not Applicable  #N/A Field Not Applicable   
2      Q3  #N/A Field Not Applicable  #N/A Field Not Applicable   
3      Q4  #N/A Field Not Applicable  #N/A Field Not Applicable   
4      Q1  #N/A Field Not Applicable  #N/A Field Not Applicable   

              GOOG UW Equity              HCP UN Equity  \
0  #N/A Field Not Applicable  #N/A Field Not Applicable   
1  #N/A Field Not Applicable  #N/A Field Not Applicable   
2  #N/A Field Not Applicable  #N/A Field Not Applicable   
3  #N/A Field Not Applicable  #N/A Field Not Applicable   
4  #N/A Field Not Applicable  #N/A Field Not Applicable   

              SYMC UW Equity               UA UN Equity  
0  #N/A Field Not Applicable  #N/A Field Not Applicable  
1  #N/A Field Not Applicable  #N/A Field Not Applicable  
2  #N/A Field Not Applicable  #N/A Field Not Applicable  
3  #N/A Field Not Applicable  #N/A Field Not Applicable  
4  #N/A Field Not Applicable  #N/A Field Not Applicable

**Observation:** These firms are all 'Object' types because there is no recorded data under them.
> This is actually missing data. From `eps_act`, there are 6 firms that are empty of data.

### This was a tricky situation to spot, but I caught it in the end. There actually is missing data after all, but we missed those 5 entire firms during the "Check Missing Data" stage, because they were recorded as objects. This means that Pandas wrongly recognized these 5 firms as "complete."

### Quality

**Missing Data**
> **In the end, I decided not to drop any rows or columns with missing data. For the sake of data preservation, I kept all missing data recorded as NaN.**

-  `eps_fc_terms` is missing the year 1999.
- `eps_act` and `eps_fc_act` have one firm with empty data: 'AMCR UN Equity'
- `eps_act` contains 7 empty firms: BRK/B UN Equity, FOX UW Equity, GOOG UW Equity, HCP UN Equity, SYMC UW Equity, UA UN Equity
--- 
- Unnormalized date formats among all DataFrames.
- Firm names across all DataFrames are capitalized and contain white space.
- Erroneous data type for `eps_act` Object columns.
- Erroneous data type for **date** under `eod_act`
- Erroneous data type for **forecast_made** under `eps_fc_terms`
- Erroneous data types for **term_forecast** under `eps_act`, `eps_fc`, and `eps_fc_terms` to DateTime quarter index
- No recorded quarterly data for `eod_act`
---
- Firm names not referenced by **firm_id**
- No recorded 20-year for each dataset.
- No recorded yearly averages for each dataset
- No recorded quarterly averages for each dataset.
---
- Features datasets 

### Tidiness
- Firm 20-year, yearly, and quarterly averages contained in different DataFrames.
- Features data is unnormalized.


### C) CLEAN <a id="clean"></a>

In [63]:
eps_fc_clean = eps_fc.copy()
eps_act_clean = eps_act.copy()
eod_act_clean = eod_act.copy()
eps_fc_terms_clean = eps_fc_terms.copy()

In [64]:
#create dictionary of all clean dfs for iteration
dict_clean = {'eps_fc_clean' : eps_fc_clean,
             'eps_act_clean' : eps_act_clean,
             'eod_act_clean' : eod_act_clean,
             'eps_fc_terms_clean' : eps_fc_terms_clean}

### Code
> **ISSUE 1:**  Unnormalized date formats among all DataFrames.

**Define:** 
- Conjoin `eps_act_clean` dates from 2 columns into 1 to match `eps_fc_clean` format.
> The format we want is YYYYQN. For example, Quarter 1 in 2005 will be 2005Q1.

- Rename both fiscal periods under `eps_act_clean` and `eps_fc_clean` to **term_forecast.**

- For both fiscal periods under `eps_fc_terms_clean`, add an underscore.
- Under `eps_fc_terms_clean`, format **term_forecasted** to YYYYQN.

In [65]:
#create new column, combine year and quarter into 1 string
eps_act_clean['term_forecast'] = eps_act_clean['Year'].map(str) + eps_act_clean['Quarter']

In [66]:
#drop 'Year' and 'Quarter'
eps_act_clean.drop(['Year', 'Quarter'], axis = 1, inplace = True)

In [67]:
#designate new column as first column
reorder_cols = eps_act_clean.columns.tolist()
reorder_cols.insert(0, reorder_cols.pop(reorder_cols.index('term_forecast')))

eps_act_clean = eps_act_clean.reindex(columns = reorder_cols)

In [68]:
#rename eps_fc_clean fiscal period to term_forecast
eps_fc_clean.rename(index = str, columns = {'Term Forecasted' : 'term_forecast'}, inplace = True)

#reassign to dictionary
dict_clean['eps_act_clean'] = eps_act_clean

In [69]:
eps_fc_clean[4:].term_forecast = '20' + eps_fc_clean[4:].term_forecast.astype(str)

D:\Anaconda3\lib\site-packages\pandas\core\generic.py:5096: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [70]:
#add an underscore for both date periods under eps_fc_terms_clean
eps_fc_terms_clean.rename(index = str, columns = {'Forecast Made' : 'forecast_made',
                                                 'Term Forecasted' : 'term_forecast'},
                         inplace = True)

In [71]:
#under term_forecaste, add a '20' before each string
eps_fc_terms_clean.term_forecast = '20' + eps_fc_terms_clean.term_forecast.astype(str)

### Test

In [72]:
#give overview of both DataFrames
eps_fc_clean.sample(3)

term_forecast  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
4         2000Q1          NaN            NaN            NaN           0.032   
15        2002Q4          NaN            NaN          0.140           0.002   
80        2019Q1        0.731          0.506          2.371           4.167   

    ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
4              NaN            NaN             NaN          0.441   
15             NaN            NaN          -0.312          0.558   
80           2.059          1.521           0.878          0.613   

    ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
4             NaN  ...          0.283           0.143          0.436   
15          0.160  ...          0.331           0.087          0.493   
80          1.855  ...          0.615           0.800          0.716   

    XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  \
4            0.138            NaN            NaN          0.138   
15           0.258            NaN            NaN          0.273   
80           0.386            0.8          0.564          0.813   

    ZBH UN Equity  ZION UW Equity  ZTS UN Equity  
4             NaN           0.668            NaN  
15          0.338           0.944            NaN  
80          1.863           1.048          0.788  

[3 rows x 506 columns]

In [73]:
eps_act_clean.sample(3)

term_forecast  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
53        2012Q2         0.73          -0.72           1.36        1.778571   
43        2009Q4         0.07          -1.03           0.37        0.402857   
17        2003Q2        -0.31          -0.47           0.40        0.002857   

    ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
53             0.8          0.820            0.02           1.09   
43             NaN          0.440           -0.23           0.98   
17             NaN          0.265           -0.25           0.16   

    ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
53           1.00  ...           0.38            0.48           3.41   
43           0.41  ...           0.37            0.28           1.27   
17           0.25  ...          -0.71            0.11           0.63   

    XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  \
53            0.57           0.92           0.48           0.71   
43            0.51           0.84            NaN           0.46   
17            0.28           0.40            NaN           0.21   

    ZBH UN Equity  ZION UW Equity  ZTS UN Equity  
53           1.22            0.30           0.35  
43           0.74           -1.26            NaN  
17           0.45            1.03            NaN  

[3 rows x 506 columns]

In [74]:
#check consistent format between both `term_forecast` fields
forecasts = pd.concat([eps_fc_clean.term_forecast, eps_act_clean.term_forecast])

#count number of instances where unique term_forecast values are the same
(forecasts.value_counts().sort_index() == 2).value_counts()

True    84
Name: term_forecast, dtype: int64

In [75]:
#check that dates are the first columns
for key, df in dict_clean.items():
    print(df.columns[:2])

Index(['term_forecast', 'A UN Equity'], dtype='object')
Index(['term_forecast', 'A UN Equity'], dtype='object')
Index(['date', 'A UN Equity'], dtype='object')
Index(['forecast_made', 'term_forecast'], dtype='object')


## The **term_forecast** fields under `eps_fc_clean`, `eps_act_clean`, and `eps_fc_terms_clean` are all normalized to YYYQN format. This means `eod_act_clean` is the only one with a different date formatting rule.

### Code
> **ISSUE 2:** Firm names across all DataFrames are capitalized and contain white space.

**Define:** Iterate across all DataFrames. Get rid of everything from the first whitespace character onward. Lowercase column names.

In [76]:
#lowercase names
for key, df in dict_clean.items():
    lower_cols = [x.lower() for x in df.columns]
    df.columns = lower_cols

In [77]:
#get rid of white space
for key, df in dict_clean.items():
    stock_symbols = [x.split(' ', 1)[0] for x in df.columns]
    df.columns = stock_symbols

### Test

In [78]:
#check that all firm names are consistent, should return 505 firm results
all_ticks = []
for key, df in dict_clean.items():
    firm_ticks = np.array(df.columns.values)
    all_ticks.append(firm_ticks)
    
#flatten array
all_ticks = np.hstack(all_ticks)

#assign tick names to Pandas Series
all_ticks = pd.Series(all_ticks)

In [79]:
#look at lowest 3 value counts, SHOULD BE DATE PERIODS
all_ticks.value_counts().sort_values()[:5]

forecast_made    1
date             1
term_forecast    3
tif              4
fdx              4
dtype: int64

In [80]:
#look at highest value counts, should be FOUR (because there are 4 dataframes all with firm names)
all_ticks.value_counts().sort_values(ascending = False)[:5]

aig    4
whr    4
maa    4
clx    4
dov    4
dtype: int64

In [81]:
#check for any upper case instance among columns
all_ticks.str.isupper().any()

False

In [82]:
#check for any white space
all_ticks.str.contains(' ').any()

False

### Code
> **ISSUE 3:** Erroneous data type for `eps_act` Object columns.

**Define:**

- Convert '#N/A Field Not Applicable' strings into 'NaN'
- Convert the 6 `eps_act_clean` Object columns into NaN type.

In [83]:
#isolate 6 firm names recorded as 'object'
obj_ticks = eps_act_clean.select_dtypes(include = 'object').columns.values

#exclude 'term_forecast'
obj_ticks = obj_ticks[1:]

In [84]:
#convert to NaN
str_replace = '#N/A Field Not Applicable'
eps_act_clean[obj_ticks] = eps_act_clean[obj_ticks].replace(str_replace, np.nan)

### Test

In [85]:
eps_act_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Columns: 506 entries, term_forecast to zts
dtypes: float64(505), object(1)
memory usage: 332.1+ KB


In [86]:
#make sure 'term_forecast' is the only object-type column
eps_act_clean.select_dtypes(include = 'object').columns.values.tolist()

['term_forecast']

### Code

> **ISSUE 4:**  Erroneous data type for **date** under `eod_act`

**Define:** Convert column `date` to DateTime object.

In [87]:
eod_act_clean.date = pd.to_datetime(eod_act_clean.date, infer_datetime_format=True)

### Test

In [88]:
eod_act_clean.sample(2)

date        a  aal     aap    aapl  abbv      abc   abmd      abt  \
21 2004-06-30  19.7280  NaN  29.453  2.3243   NaN  14.5017  12.58  19.5026   
20 2004-03-31  21.3114  NaN  27.113  1.9321   NaN  13.2645   8.10  18.3939   

      acn  ...  xel   xlnx    xom    xray      xrx  xyl      yum    zbh  \
21  27.48  ...  NaN  33.31  44.41  26.050  38.2016  NaN  13.3816  88.20   
20  24.80  ...  NaN  38.00  41.59  22.165  38.3860  NaN  13.6584  73.78   

     zion  zts  
21  61.45  NaN  
20  57.20  NaN  

[2 rows x 506 columns]

In [89]:
#confirm datetime64 present
eod_act_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Columns: 506 entries, date to zts
dtypes: datetime64[ns](1), float64(505)
memory usage: 332.1 KB


### Code
> **ISSUE 5:** Erroneous data type for **forecast_made** under `eps_fc_terms`.

**Define:**

- Convert column **forecast_made** to DateTime object.

In [90]:
eps_fc_terms_clean.forecast_made = pd.to_datetime(eps_fc_terms_clean.forecast_made, infer_datetime_format=True)

### Test

In [91]:
#confirm presence of datetime64 object
eps_fc_terms_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 80 entries, 0 to 79
Columns: 507 entries, forecast_made to zts
dtypes: datetime64[ns](1), float64(505), object(1)
memory usage: 317.5+ KB


In [92]:
eps_fc_terms_clean.sample(2)

forecast_made term_forecast      a    aal    aap   aapl  abbv    abc  \
50    2012-04-01        2012Q3  0.832  0.135  1.545  1.396   NaN  0.703   
2     2000-04-01        2000Q3    NaN    NaN    NaN  0.029   NaN    NaN   

     abmd    abt  ...    xel   xlnx    xom   xray    xrx    xyl    yum    zbh  \
50 -0.007  1.265  ...  0.699  0.366  2.048  0.560  1.105  0.498  0.978  1.158   
2     NaN  0.414  ...  0.840  0.189  0.425  0.149    NaN    NaN  0.218    NaN   

     zion  zts  
50  0.398  NaN  
2   0.803  NaN  

[2 rows x 507 columns]

### Code
> **ISSUE 6:** Erroneous data types for **term_forecast** under `eps_act`, `eps_fc`, and `eps_fc_terms` to DateTime quarter index

**Define:** Convert YYYYQQ formats into DateTime quarter index type

In [93]:
#helper function to turn a column into Quarter Period type
def convert_period(df, col):
    return pd.to_datetime(df[col]).dt.to_period('Q')
# eps_fc_terms_clean.term_forecast = pd.PeriodIndex(eps_fc_terms_clean.term_forecast, freq='Q')

In [94]:
#convert term_forecast fields to Period type
eps_fc_clean.term_forecast = convert_period(eps_fc_clean, 'term_forecast')
eps_act_clean.term_forecast = convert_period(eps_act_clean, 'term_forecast')
eps_fc_terms_clean.term_forecast = convert_period(eps_fc_terms_clean, 'term_forecast')

### Test

In [95]:
#verify Period Object types
print(eps_fc_clean.dtypes['term_forecast'])
print(eps_act_clean.dtypes['term_forecast'])
print(eps_fc_terms_clean.dtypes['term_forecast'])

period[Q-DEC]
period[Q-DEC]
period[Q-DEC]


### Code
> **ISSUE 7:** No recorded quarterly data for `eod_act`

**Define:** Add a new column **term_forecast** denoting *year* and *quarter*

In [96]:
eod_act_clean['term_forecast'] = eod_act_clean.date.dt.to_period('Q')

In [97]:
#reorder columns
eod_cols = ['date', 'term_forecast'] + [col for col in eod_act_clean.iloc[:, 1:] if col!='term_forecast']

In [98]:
eod_act_clean = eod_act_clean.reindex(columns = eod_cols)

### Test

In [99]:
#verify column order
eod_act_clean.head()

date term_forecast        a  aal  aap    aapl  abbv     abc    abmd  \
0 1999-03-31        1999Q1      NaN  NaN  NaN  1.2835   NaN  8.2934   6.250   
1 1999-06-30        1999Q2      NaN  NaN  NaN  1.6540   NaN  6.1859   6.875   
2 1999-09-30        1999Q3      NaN  NaN  NaN  2.2612   NaN  5.7462   7.750   
3 1999-12-31        1999Q4  52.0909  NaN  NaN  3.6719   NaN  3.6843  18.375   
4 2000-03-31        2000Q1  70.0721  NaN  NaN  4.8504   NaN  3.6388  20.250   

       abt  ...  xel     xlnx      xom    xray       xrx  xyl      yum  zbh  \
0  20.9504  ...  NaN  20.2813  35.2813  7.7500  140.6214  NaN  12.6284  NaN   
1  20.3630  ...  NaN  28.6250  38.5625  9.6250  155.6057  NaN   9.7297  NaN   
2  16.4471  ...  NaN  32.7656  37.9688  7.5833  110.4883  NaN   7.3591  NaN   
3  16.2513  ...  NaN  45.4688  40.2813  7.8750   59.7723  NaN   6.9434  NaN   
4  15.7478  ...  NaN  82.8125  38.9063  9.4583   68.4994  NaN   5.5839  NaN   

      zion  zts  
0  66.5000  NaN  
1  63.5000  NaN  
2  55.1250  NaN  
3  59.1875  NaN  
4  41.6250  NaN  

[5 rows x 507 columns]

In [100]:
#verify 84 columns
eod_act_clean.shape[0]

84

In [101]:
eod_act_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Columns: 507 entries, date to zts
dtypes: datetime64[ns](1), float64(505), period[Q-DEC](1)
memory usage: 332.8 KB


## All full years are normalized to YYYY-MM-DD, and all quarterly records to YYYY-QQ.

## This is to enable for more efficient handling, cleaning, and classifying of data later on.

### Code

**ISSUE 7:** Firm names not referenced by **firm_id**

**Define:**
- Assign a **firm_id** to each firm for future normalization.

- Generate new CSV named `firms.csv`

In [102]:
#get firm names
firm_names = eps_act_clean.columns[1:].to_numpy()

In [103]:
#assign ids to firm names
firm_names_ids = {}
counter = 0

for firm in firm_names:
    firm_names_ids.update({firm: counter})
    counter += 1

In [104]:
#generate new DF
firm_ids = pd.DataFrame(list(firm_names_ids.items())).rename(columns = {0: 'firm', 1: 'firm_id'})

In [105]:
#switch firm and firm_id columns
firm_ids = firm_ids.reindex(columns = ['firm_id', 'firm'])

### Test

In [106]:
firm_ids.sample(10)

firm_id  firm
113      113   coo
463      463   usb
348      348  ntap
32        32  amcr
344      344   noc
99        99    ci
270      270  keys
265      265  jnpr
454      454   ual
132      132   dal

In [107]:
firm_ids.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 505 entries, 0 to 504
Data columns (total 2 columns):
firm_id    505 non-null int64
firm       505 non-null object
dtypes: int64(1), object(1)
memory usage: 8.0+ KB


### Code
> **ISSUE 7:** No recorded 20-year averages for each dataset.

**Define:**


- Isolate 20-year averages for each firm into its own DataFrame

- Create new DataFrame `twenty_avgs` depicting all 20-year averages for each firm

In [114]:
#convert foreign keys to str temporarily
firm_ids_str = {str(k):str(v) for k, v in firm_names_ids.items()}
# firm_ids_str = {y:x for x, y in firm_ids_str.items()}

In [115]:
#helper function to get 20-year averages
def get_twenty_yr_avg(df, index_names, col_name, firm_ids_str):
    df = pd.DataFrame({col_name : df[index_names].mean()}).rename_axis('firm_id').reset_index()
    df['firm_id'] = df['firm_id'].map(firm_ids_str)
    return df

In [116]:
#get 20-year averages
eps_fc_twenty_avg = get_twenty_yr_avg(eps_fc_clean, firm_names, 'eps_fc', firm_ids_str)
eps_act_twenty_avg = get_twenty_yr_avg(eps_act_clean, firm_names, 'eps_act', firm_ids_str)
eod_act_twenty_avg = get_twenty_yr_avg(eod_act_clean, firm_names, 'eod_act', firm_ids_str)
eps_fc_terms_twenty_avg = get_twenty_yr_avg(eps_fc_terms_clean, firm_names, 'eps_fc_terms', firm_ids_str)

In [118]:
#put averages into a list
twenty_avg_list = [eps_fc_twenty_avg,
                  eps_act_twenty_avg,
                  eod_act_twenty_avg,
                  eps_fc_terms_twenty_avg]

In [119]:
#merge all Series
df_twenty_year_avgs = reduce(lambda x, y: pd.merge(x, y, on = 'firm_id', how = 'outer'), twenty_avg_list)

### Test

In [120]:
df_twenty_year_avgs.sample(5)

firm_id    eps_fc   eps_act     eod_act  eps_fc_terms
463     463  0.610083  0.602996   32.690114      0.650150
214     214  2.869262  3.255181  145.443341      3.472923
466     466  0.625107  0.568186   52.221702      0.681850
222     222  0.755012  0.783893   70.500000      0.784138
401     401  1.000405  1.030357   73.834795      1.055000

In [124]:
#check there are 505 firms
df_twenty_year_avgs.firm_id.nunique()

505

In [125]:
#check missing data
df_twenty_year_avgs.isna().any()

firm_id         False
eps_fc          False
eps_act          True
eod_act         False
eps_fc_terms     True
dtype: bool

In [126]:
#check duplicate data
df_twenty_year_avgs.duplicated().any()

False

### Code
> **ISSUE 8:** No recorded yearly averages for each dataset.

**Define:**
- Create 4 separate DataFrames for all attributes
- Rename columns to "Feature_Year" (e.g. eps_fc_1999, eod_act_2000, etc.)
- Outer merge all DataFrames to create new DataFrame, `df_yearly_avgs`, on **firms** attribute

In [165]:
#helper function to calculate yearly average of each DataFrame
def calculate_df_yearly_avgs(df, col, firm_ids_str):
    #extract year from dates, reset index, and remove index name
    df = df.groupby(df[col].dt.year).mean().transpose().rename_axis('firm_id').reset_index().rename_axis(None, axis = 1)
    
    df['firm_id'] = df['firm_id'].map(firm_ids_str)
    
    return df

In [166]:
#helper function to melt columns
def melt_avgs(df, firm_id, var_name):
    return df.melt(id_vars = [firm_id],
                value_vars = df.columns.difference([firm_id]),
                var_name = var_name,
                value_name = 'average')

In [228]:
#calculate and melt
def create_df_yearly_avgs(df, col, firm_ids_str, firm_id):
    df = calculate_df_yearly_avgs(df, 'term_forecast', firm_ids_str)
    df = melt_avgs(df, firm_id, 'year')
    return df

In [230]:
#generate DataFrames of all yearly averages
eps_fc_yearly_avgs = create_df_yearly_avgs(eps_fc_clean, 'term_forecast', firm_ids_str, 'firm_id')
eps_act_yearly_avgs = create_df_yearly_avgs(eps_act_clean, 'term_forecast', firm_ids_str, 'firm_id')

#for eod_act, doesn't matter if we use date or term_forecast field
eod_act_yearly_avgs = create_df_yearly_avgs(eod_act_clean, 'date', firm_ids_str, 'firm_id')
eps_fc_terms_yearly_avgs = create_df_yearly_avgs(eps_fc_terms_clean, 'term_forecast', firm_ids_str, 'firm_id')

In [231]:
#put DataFrames in a dictionary
yearly_avgs_dict = {'eps_fc_yearly_avgs': eps_fc_yearly_avgs,
                   'eps_act_yearly_avgs': eps_act_yearly_avgs,
                   'eod_act_yearly_avgs' : eod_act_yearly_avgs,
                   'eps_fc_terms_yearly_avgs' : eps_fc_terms_yearly_avgs}

In [233]:
def store_df(path, df, csv_name):
    df.to_csv(path + csv_name, encoding = 'utf-8', index = False)

In [234]:
#store yearly averages
store_df(PATH_CLEAN_AVGS_SUB, eps_fc_yearly_avgs, 'yearly-avgs-eps-fc.csv')
store_df(PATH_CLEAN_AVGS_SUB, eps_act_yearly_avgs, 'yearly-avgs-eps-act.csv')
store_df(PATH_CLEAN_AVGS_SUB, eod_act_yearly_avgs, 'yearly-avgs-eod-act.csv')
store_df(PATH_CLEAN_AVGS_SUB, eps_fc_terms_yearly_avgs, 'yearly-avgs-eps-fc-terms.csv')

In [235]:
#add new column entailing "type" of feature before concat
eps_fc_yearly_avgs = eps_fc_yearly_avgs.assign(feature = 'eps_fc')
eps_act_yearly_avgs = eps_act_yearly_avgs.assign(feature = 'eps_act')
eod_act_yearly_avgs = eod_act_yearly_avgs.assign(feature = 'eod_act')
eps_fc_terms_yearly_avgs = eps_fc_terms_yearly_avgs.assign(feature = 'eps_fc_terms')

In [236]:
#merge all yearly average DataFrames
df_yearly_avgs = pd.concat(yearly_avgs_arr)

### Test

In [237]:
#each average, individually

#eps_fc_yearly
eps_fc_yearly_avgs.sample(10)

firm_id  year  average feature
907      402  2000      NaN  eps_fc
4838     293  2008  0.32750  eps_fc
6356     296  2011  0.11175  eps_fc
3734     199  2006      NaN  eps_fc
5394     344  2009  1.20400  eps_fc
5001     456  2008  0.42275  eps_fc
4038     503  2006  1.39200  eps_fc
9610      15  2018  0.74875  eps_fc
1308     298  2001      NaN  eps_fc
6733     168  2012  0.42400  eps_fc

In [238]:
#eps_fc_yearly
eps_act_yearly_avgs.sample(10)

firm_id  year   average  feature
589        84  2000       NaN  eps_act
10340     240  2019  2.166667  eps_act
10403     303  2019  0.163333  eps_act
2160      140  2003  0.142500  eps_act
8081        1  2015  2.902500  eps_act
4278      238  2007  0.240000  eps_act
9529      439  2017  0.437500  eps_act
6255      195  2011       NaN  eps_act
9662       67  2018  5.410000  eps_act
7487      417  2013  0.405000  eps_act

In [239]:
#eod_act_yearly
eod_act_yearly_avgs.sample(10)

firm_id  year     average  feature
9377     287  2017  123.420000  eod_act
3796     261  2006   20.360275  eod_act
8563     483  2015   69.685000  eod_act
9519     429  2017  141.822500  eod_act
7945     370  2014   25.462500  eod_act
8703     118  2016   49.052500  eod_act
3924     389  2006         NaN  eod_act
6016     461  2010   65.142500  eod_act
8300     220  2015   77.735000  eod_act
585       80  2000   16.195250  eod_act

In [240]:
#eps_fc_terms_yearly
eps_fc_terms_yearly_avgs.sample(10)

firm_id  year  average       feature
3731     196  2007      NaN  eps_fc_terms
5710     155  2011  0.09075  eps_fc_terms
7826     251  2015  0.28400  eps_fc_terms
1378     368  2002  0.55100  eps_fc_terms
893      388  2001  0.43450  eps_fc_terms
8814     229  2017  0.49675  eps_fc_terms
2779     254  2005  0.70975  eps_fc_terms
7074       4  2014  0.78000  eps_fc_terms
9538     448  2018  0.63800  eps_fc_terms
1981     466  2003  0.20050  eps_fc_terms

In [241]:
#verify all columns are consistent
print(eps_fc_yearly_avgs.columns, '\n',
      eps_act_yearly_avgs.columns, '\n',
      eod_act_yearly_avgs.columns, '\n',
      eps_fc_terms_yearly_avgs.columns)

Index(['firm_id', 'year', 'average', 'feature'], dtype='object') 
 Index(['firm_id', 'year', 'average', 'feature'], dtype='object') 
 Index(['firm_id', 'year', 'average', 'feature'], dtype='object') 
 Index(['firm_id', 'year', 'average', 'feature'], dtype='object')


In [242]:
#verify there are 505 firms in each dataframe
print(eps_fc_yearly_avgs.firm_id.nunique(), '\n',
      eps_act_yearly_avgs.firm_id.nunique(), '\n',
      eod_act_yearly_avgs.firm_id.nunique(), '\n',
      eps_fc_terms_yearly_avgs.firm_id.nunique())

505 
 505 
 505 
 505


In [243]:
#all yearly averages
df_yearly_avgs.sample(10)

firm_id  year    average       feature
4613      68  2009    0.67700  eps_fc_terms
75        75  1999   41.38500       eod_act
7891     316  2014   61.71250       eod_act
5931     376  2010   53.13750       eod_act
6569       4  2012        NaN        eps_fc
2408     388  2003    0.87750       eps_act
5724     169  2010   -0.08725        eps_fc
9115      25  2017  118.84500       eod_act
1720     205  2002        NaN        eps_fc
8642      57  2017    1.41825  eps_fc_terms

In [244]:
#verify 505 firms in all yearly averages df
df_yearly_avgs.firm_id.nunique()

505

### Code

> **ISSUE 9:** No recorded quarterly averages for each dataset.

**Define:** 

- Parse `eod_act_clean` dates  by **calendar quarter average** with quarterly data in a new DataFrame.
- Create separate DataFrames containing quarterly averages for `eps_fc_clean`, `eps_act_clean`, and `eps_fc_terms_clean` 
- Rename columns to "Quarter_Year_Feature" (e.g. **q1_eps_fc,** etc.)
- Outer merge all DataFrames into a new df `df_quarterly_avgs` on **firms**

In [245]:
#helper function to calculate quarterly average of each DataFrame
def calculate_df_quarterly_avgs(df, col, firm_ids_str):
    #extract quarter from Period objects
    df =  df.groupby(df[col].dt.strftime('Q%q')).mean().transpose().rename_axis('firm_id').reset_index().rename_axis(None, axis = 1)
    df['firm_id'] = df['firm_id'].map(firm_ids_str)
    
    return df

In [246]:
#calculate and melt
def create_df_quarterly_avgs(df, col, firm_ids_str, firm_id):
    df = calculate_df_quarterly_avgs(df, 'term_forecast', firm_ids_str)
    df = melt_avgs(df, firm_id, 'quarter')
    
    #lowercase quarter column
    df.quarter = df.quarter.str.lower()
    return df

In [248]:
#generate all quarterly average DataFrames
eps_fc_quarter_avgs = create_df_quarterly_avgs(eps_fc_clean, 'term_forecast', firm_ids_str, 'firm_id')
eps_act_quarter_avgs = create_df_quarterly_avgs(eps_act_clean, 'term_forecast', firm_ids_str, 'firm_id')
eod_act_quarter_avgs = create_df_quarterly_avgs(eod_act_clean, 'term_forecast', firm_ids_str, 'firm_id')
eps_fc_terms_quarter_avgs = create_df_quarterly_avgs(eps_fc_terms_clean, 'term_forecast', firm_ids_str, 'firm_id')

In [249]:
#store quarterly averages
store_df(PATH_CLEAN_AVGS_SUB, eps_fc_quarter_avgs, 'quarter-avgs-eps-fc.csv')
store_df(PATH_CLEAN_AVGS_SUB, eps_act_quarter_avgs, 'quarter-avgs-eps-act.csv')
store_df(PATH_CLEAN_AVGS_SUB, eod_act_quarter_avgs, 'quarter-avgs-eod-act.csv')
store_df(PATH_CLEAN_AVGS_SUB, eps_fc_terms_quarter_avgs, 'quarter-avgs-eps-fc-terms.csv')

In [250]:
#add new column entailing "type" of feature before concat
eps_fc_quarter_avgs = eps_fc_quarter_avgs.assign(feature = 'eps_fc')
eps_act_quarter_avgs = eps_act_quarter_avgs.assign(feature = 'eps_act')
eod_act_quarter_avgs = eod_act_quarter_avgs.assign(feature = 'eod_act')
eps_fc_terms_quarter_avgs = eps_fc_terms_quarter_avgs.assign(feature = 'eps_fc_terms')

In [251]:
#put DataFrames in array
quarterly_avgs_arr = [eps_fc_quarter_avgs,
                  eps_act_quarter_avgs,
                  eod_act_quarter_avgs,
                  eps_fc_terms_quarter_avgs]

In [255]:
#merge all quarterly average DataFrames
df_quarter_avgs = pd.concat(quarterly_avgs_arr)

### Test

In [256]:
#each quarter individually

#eps_fc_quarter
eps_fc_quarter_avgs.sample(10)

firm_id quarter   average feature
187      187      q1  0.568111  eps_fc
949      444      q2  0.709143  eps_fc
460      460      q1  0.656000  eps_fc
882      377      q2  1.410524  eps_fc
528       23      q2  0.529000  eps_fc
1310     300      q3  0.192667  eps_fc
115      115      q1  0.700095  eps_fc
547       42      q2  1.997000  eps_fc
1429     419      q3  0.874800  eps_fc
384      384      q1  0.977286  eps_fc

In [258]:
#eps_act_quarter
eps_act_quarter_avgs.sample(10)

firm_id quarter   average  feature
1646     131      q4  0.409739  eps_act
1479     469      q3  1.233662  eps_act
1057      47      q3  0.391518  eps_act
1124     114      q3  1.362619  eps_act
1859     344      q4  1.326955  eps_act
1579      64      q4  0.762233  eps_act
113      113      q1  0.587857  eps_act
1183     173      q3  0.340500  eps_act
458      458      q1  1.090417  eps_act
1480     470      q3  0.822857  eps_act

In [259]:
#eod_act_quarter
eod_act_quarter_avgs.sample(10)

firm_id quarter     average  feature
1928     413      q4   68.631100  eod_act
662      157      q2   67.066848  eod_act
2008     493      q4   91.902941  eod_act
117      117      q1   37.125119  eod_act
929      424      q2   34.567273  eod_act
1570      55      q4   57.624500  eod_act
1528      13      q4  120.747222  eod_act
58        58      q1  310.794524  eod_act
173      173      q1   42.893900  eod_act
604       99      q2   67.613024  eod_act

In [260]:
#eps_fc_terms_quarter
eps_fc_terms_quarter_avgs.sample(10)

firm_id quarter   average       feature
323      323      q1  0.713000  eps_fc_terms
981      476      q2  0.783250  eps_fc_terms
1932     417      q4  0.531929  eps_fc_terms
1138     128      q3  0.775150  eps_fc_terms
1550      35      q4  2.445250  eps_fc_terms
1447     437      q3  0.582250  eps_fc_terms
259      259      q1  0.466250  eps_fc_terms
941      436      q2  1.295200  eps_fc_terms
1851     336      q4  0.573571  eps_fc_terms
1509     499      q3  0.983643  eps_fc_terms

In [262]:
#verify that all columns are consistent
print(eps_fc_quarter_avgs.columns, '\n',
      eps_act_quarter_avgs.columns, '\n',
      eod_act_quarter_avgs.columns, '\n',
      eps_fc_terms_quarter_avgs.columns)

Index(['firm_id', 'quarter', 'average', 'feature'], dtype='object') 
 Index(['firm_id', 'quarter', 'average', 'feature'], dtype='object') 
 Index(['firm_id', 'quarter', 'average', 'feature'], dtype='object') 
 Index(['firm_id', 'quarter', 'average', 'feature'], dtype='object')


In [264]:
#verify there are 505 firms in each dataframe
print(eps_fc_quarter_avgs.firm_id.nunique(), '\n',
      eps_act_quarter_avgs.firm_id.nunique(), '\n',
      eod_act_quarter_avgs.firm_id.nunique(), '\n',
      eps_fc_terms_quarter_avgs.firm_id.nunique())

505 
 505 
 505 
 505


In [266]:
#all quarterly averages
df_quarter_avgs.sample(10)

firm_id quarter    average       feature
1635     120      q4   0.042052       eps_act
643      138      q2   0.359900  eps_fc_terms
1704     189      q4   0.250044       eps_act
491      491      q1   0.351538       eps_act
1384     374      q3   0.415667        eps_fc
433      433      q1  48.890524       eod_act
52        52      q1   0.161500  eps_fc_terms
1172     162      q3   1.100050  eps_fc_terms
1487     477      q3  37.657443       eod_act
196      196      q1   0.847000  eps_fc_terms

In [267]:
#verify 505 firms in df_quarter_avgs
df_quarter_avgs.firm_id.nunique()

505

### Code
>**ISSUE 10:** Firm 20-year, yearly, and quarterly averages contained in different DataFrames.

**Define:** 
- Outer merge `df_twenty_year_avgs`, `df_yearly_avgs`, and `df_quarter_avgs` on **firm** column.

In [206]:
#put all average DFs in an array
avgs_array = [df_twenty_year_avgs, df_yearly_avgs, df_quarter_avgs]

In [207]:
df_all_avgs = pd.concat(avgs_array, join='outer', axis=1).fillna(method='ffill')

### Test

In [209]:
df_all_avgs.head()

firm    eps_fc   eps_act    eod_act  eps_fc_terms  1999_eps_fc  2000_eps_fc  \
a     0.531678  0.352169  32.896718      0.567579          NaN          NaN   
aal   0.339879 -0.388520  40.303125      0.709745          NaN          NaN   
aap   1.051973  0.996901  76.856583      1.163366          NaN          NaN   
aapl  0.873143  0.944501  55.145230      0.933613        0.021       0.0285   
abbv  1.319607  0.778921  66.260714      1.322593        0.021       0.0285   

firm  2001_eps_fc  2002_eps_fc  2003_eps_fc  2004_eps_fc  ...  q3_eps_act  \
a             NaN          NaN          NaN          NaN  ...    0.191000   
aal           NaN          NaN          NaN          NaN  ...    0.071912   
aap       0.02850      0.21075       0.3185        0.414  ...    1.007222   
aapl     -0.00575      0.00575       0.0025        0.010  ...    0.750119   
abbv     -0.00575      0.00575       0.0025        0.010  ...    0.968750   

firm  q4_eps_act  q1_eod_act  q2_eod_act  q3_eod_act  q4_eod_act  \
a       0.416667   34.058320   32.998630   31.991835   32.538085   
aal    -0.871865   42.735000   38.682500   37.785000   42.010000   
aap     0.670000   77.747000   76.724222   78.780333   74.174778   
aapl    0.878384   53.321171   54.858752   60.959586   51.256220   
abbv    0.249505   64.034286   66.394286   68.447143   66.167143   

firm  q1_eps_fc_terms  q2_eps_fc_terms  q3_eps_fc_terms  q4_eps_fc_terms  
a            0.542500         0.537857         0.558000         0.627667  
aal          0.048769         1.261429         1.159643         0.321929  
aap          1.453235         1.262778         1.177556         0.776000  
aapl         1.231100         0.912950         0.750250         0.840150  
abbv         1.225000         1.293143         1.359429         1.398857  

[5 rows x 103 columns]

In [210]:
#verify 103 columns
df_all_avgs.shape

(505, 103)

# III) Store Data <a id = "store"></a>

**Disclaimer:** I decided to put the following CSVs under the ***./data/clean/averages/components*** directory because they form *part* of the main CSVs under the ***./data/clean/averages*** directory:
- `quarter-avgs-eod-act.csv`
- `quarter-avgs-eps-act.csv`
- `quarter-avgs-eps-fc.csv`
- `quarter-avgs-eps-fc-terms.csv`

> The above are all components of the `quarter-avgs.csv` under **/averages/**

- `yearly-avgs-eod-act.csv`
- `yearly-avgs-eps-act.csv`
- `yearly-avgs-eps-fc.csv`
- `yearly-avgs-eps-fc-terms.csv`

> The above are all components of the `yearly-avgs.csv` under **/averages/**

**Additionally, I decided to keep `avgs.csv` under the *./data/clean/averages/* directory.** Although this CSV containing all averages would be highly inefficient for the data exploration stage, I decided it would be best to keep it on hand for convenience.

Using `avgs.csv` would mean having the parse through the column names to isolate 20-year, yearly, and quarterly averages, but all of that redundance can be avoided by simply focusing on the `twenty-year-avgs`, `yearly-avgs`, and `quarter-avgs` CSVs alone.

## Firms

In [545]:
firm_ids.to_csv(PATH_CLEAN + 'firms.csv', encoding = 'utf-8', index = False)

## Averages

In [268]:
#twenty-year-averages
df_twenty_year_avgs.to_csv(PATH_CLEAN_AVGS + 'twenty-year-avgs.csv', encoding = 'utf-8', index = False)

In [269]:
#yearly averagas
df_yearly_avgs.to_csv(PATH_CLEAN_AVGS + 'yearly-avgs.csv', encoding = 'utf-8', index = False)

In [270]:
#quarterly averages
df_quarter_avgs.to_csv(PATH_CLEAN_AVGS + 'quarter-avgs.csv', encoding = 'utf-8', index = False)

In [217]:
#all averages
df_all_avgs.to_csv(PATH_CLEAN_AVGS + 'avgs.csv', encoding = 'utf-8', index = False)

## Clean Standalone Datasets

In [218]:
eps_fc_clean.to_csv(PATH_CLEAN + 'eps-fc.csv')
eps_act_clean.to_csv(PATH_CLEAN + 'eps-act.csv')
eod_act_clean.to_csv(PATH_CLEAN + 'eod-act.csv')
eps_fc_terms_clean.to_csv(PATH_CLEAN + 'eps-fc-terms.csv')

# IV) Notes



- When dealing with the CSVs depicting averages, we need to keep in mind **missing data** in our interpretations.
- I decided to put the years as columns under yearly average DF generation because: the amount of columns generated here, TK, would be less than the amount of 
	firms there are, 505. [TK link to a study showing how having too many columns take up memory vs rows]
    
- keep in mind that `eod_act` is based on **calendar years** not fiscal years.

In [136]:

#convert notebook to HTML